object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java

```diff
@@ -279,6 +306,276 @@ private static String getExactStringValue(Tree tree, ValueAnnotatedTypeFactory f
+      worklist.removeLast();
+      List<Node> nodes = getBlockNodes(state.first);
+      Set<AssignmentNode> newDefs = new HashSet<>(state.second);
+
+      for (Node node : nodes) {
+        if (node instanceof AssignmentNode) {
+          Node lhs = ((AssignmentNode) node).getTarget();
+          Node rhs = ((AssignmentNode) node).getExpression();
+
+          if (lhs instanceof LocalVariableNode && hasAlwaysCall(lhs.getType())) {
+
+            // Reassignment to the lhs
+            if (isVarInPrevDefs(newDefs, (LocalVariableNode) lhs)) {
+              AssignmentNode latestAssignmentNode =
+                  getAssignmentNodeOfVar(newDefs, (LocalVariableNode) lhs);
+              reportAlwaysCallErrors(latestAssignmentNode, getStoreBefore(node), null);
+              newDefs.remove(latestAssignmentNode);
+            }
+
+            // If the rhs is ObjectCreationNode, MethodInvocationNode, MethodAccessNode or
+            // LocalVariableNode that exists in the newDefs (Note that if the localVariableNode
+            // exists in the newDefs it means it is assigned to a non-null values), then it adds
+            // the the AssignmentNode to the newDefs.
+            if ((rhs instanceof ObjectCreationNode)
+                || (rhs instanceof MethodInvocationNode)
+                || (rhs instanceof MethodAccessNode)
+                || (rhs instanceof LocalVariableNode
+                    && isVarInPrevDefs(newDefs, (LocalVariableNode) rhs))) {
+              newDefs.add((AssignmentNode) node);
+            }
+
+            // Ownership Transfer
+            if (rhs instanceof LocalVariableNode
+                && isVarInPrevDefs(newDefs, (LocalVariableNode) rhs)) {
+              newDefs.remove(getAssignmentNodeOfVar(newDefs, (LocalVariableNode) rhs));
+            }
+          }
+        }
+
+        // Remove the returned localVariableNode from newDefs.
+        if (node instanceof ReturnNode) {
+          Node result = ((ReturnNode) node).getResult();
+          if (result instanceof LocalVariableNode
+              && isVarInPrevDefs(newDefs, (LocalVariableNode) result)) {
+            newDefs.remove(getAssignmentNodeOfVar(newDefs, (LocalVariableNode) result));
+          }
+        }
+      }
+
+      for (BlockImpl succ : getSuccessors(state.first)) {
+        Set<AssignmentNode> toRemove = new HashSet<>();
+
+        // There is nothing needs to be checked
+        if (newDefs.size() == 0) {
+          propagate(Pair.of(succ, newDefs), visited, worklist);
+          continue;
+        }
+
+        CFStore succRegularStore = getAnalysis().getInput(succ).getRegularStore();
+
+        for (AssignmentNode assignmentNode : newDefs) {
+
+          // If the successor block is the exit block or if the set of the pair of block id and set
+          // of defs is repeated in a loop or if the variable is going out of scope
+          if (succ instanceof SpecialBlockImpl
+              || visited.contains(Pair.of(succ.getId(), newDefs))
```

Comment:
What happens if you remove this case?  I don't understand why it is needed

Reply:
Then it can't find errors inside the loops properly:
void overWrittingVarInLoop() {
        // :: error: missing.alwayscall
        Foo f = new Foo();
        while (true) {
            // :: error: missing.alwayscall
            f = new Foo();
        }
    }

The other example is:
    void loopWithNestedBranches(boolean b) {
        Foo f = null;
        while (true) {
            if (b) {
                // :: error: missing.alwayscall
                f = new Foo();
            } else {
                // :: error: missing.alwayscall
                f = new Foo();
            }
        }
    }

It doesn't report an error for the if branch! But it reports an error for the then branch because it's checked first and then the if branch overwrites it!